internal/consensus/state.go

```diff
@@ -1314,6 +1314,12 @@ func (cs *State) defaultDoPrevote(height int64, round int32) {
 		return
 	}
 
+	if !cs.Proposal.Timestamp.Equal(cs.ProposalBlock.Header.Time) {
+		logger.Debug("proposal timestamp not equal, prevoting nil")
```

Comment:
> should we not include this checking in the ValidateBlock logic?

This is a good question. I've been going back and forth on this trying to understand what the right way to shim this new information into that method would be. It means that one of three things would have to happen:

1. `state.State` would need to have the proposal as a field. State is currently composed of all fields that can be determined from the previous block itself. 
2. `ValidateBlock` would need to take a proposal as a parameter. 
3. `BlockExecutor` would need a reference to the proposal.

I'm not convinced these are great ideas. For 1., the `State` is currently made up of only things that can be determined from previous blocks, so we'd be changing the purpose of that type.

For 2., we use [ValidateBlock when replaying old blocks](https://github.com/tendermint/tendermint/blob/88ba6c70fcd624ecb006c002a7a85ff3b037296e/internal/consensus/replay.go#L517). This would mean keeping old proposals as part of block verification for old blocks. I'm not sure this is a meaningful thing to do, but maybe I don't see how it is yet.

Finally, 3. Is a similar issue to 2, with the added difficulty that block executor would need some concurrency safe way to access the current round's proposal. 

I think overall that 1. would make the most sense, but I'm not certain it's necessary. Let me know if you disagree or if I've missed something.